File: backtracking/permutations2/permutations_2.py
from typing import List
import copy

class Solution:

    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        
        def backtrack(candidates: List[int], solution: List[int]):

            if not candidates:
                solutions.add(tuple(copy.copy(solution)))
                return

            for ind, candidate in enumerate(candidates):
                solution.append(candidate)
                candidates_copy = copy.copy(candidates)
                candidates_copy.pop(ind)
                backtrack(candidates_copy, solution)
                solution.pop(-1)
        
        solutions = set()
        if nums:
            backtrack(candidates=nums, solution=[])
        
        final_solutions = [list(solution) for solution in solutions]

        return final_solutions


if __name__ == "__main__":
    print(Solution().permuteUnique([1, 1, 2]))

File: backtracking/permutations2/permutations_2.go
package permutations2

func backTrack(solutions *[][]int, candidates []int, solution []int, freqMap map[int]int, uniqueNums []int, elementsNum int) {

	if len(solution) == elementsNum {
		solutionCopy := make([]int, len(solution))
		copy(solutionCopy, solution)
		*solutions = append(*solutions, solutionCopy)
		return
	}

	for _, candidate := range uniqueNums {
		if freqMap[candidate] > 0 {
			freqMap[candidate]--
			backTrack(solutions, candidates, append(solution, candidate), freqMap, uniqueNums, elementsNum)
			freqMap[candidate]++
		}
	}
}

// PermuteUnique ...
func PermuteUnique(nums []int) [][]int {

	var solutions [][]int
	freqMap := make(map[int]int)
	uniqueNums := make([]int, 0)

	for _, num := range nums {
		if _, ok := freqMap[num]; !ok {
			uniqueNums = append(uniqueNums, num)
			freqMap[num] = 1
		} else {
			freqMap[num]++
		}
	}

	backTrack(&solutions, nums, []int{}, freqMap, uniqueNums, len(nums))

	return solutions
}
